src/app/BufferedReadCallback.cpp

```diff
@@ -0,0 +1,239 @@
+
+        reader.Init(std::move(bufHandle));
+
+        err = reader.Next();
+        SuccessOrExit(err);
+
+        err = writer.CopyElement(TLV::AnonymousTag, reader);
+        SuccessOrExit(err);
+    }
+
+    err = writer.EndContainer(outerType);
+    SuccessOrExit(err);
+
+    writer.Finalize(&handle);
+    aReader.Init(std::move(handle));
+
+exit:
+    return err;
+}
+
+CHIP_ERROR BufferedReadCallback::BufferData(const ConcreteAttributePath & aPath, TLV::TLVReader * apData)
+{
+    CHIP_ERROR err;
+
+    if (aPath.mListOp == ConcreteAttributePath::ListOperation::ReplaceAll)
+    {
+        TLV::TLVType outerContainer;
+
+        VerifyOrExit(apData->GetType() == TLV::kTLVType_Array, err = CHIP_ERROR_INVALID_TLV_ELEMENT);
+        mBufferedList.clear();
+
+        err = apData->EnterContainer(outerContainer);
+        SuccessOrExit(err);
+
+        while ((err = apData->Next()) == CHIP_NO_ERROR)
+        {
+            System::PacketBufferTLVWriter writer;
+            System::PacketBufferHandle handle = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLengthBytes);
+
+            writer.Init(std::move(handle), false);
+
+            err = writer.CopyElement(TLV::AnonymousTag, *apData);
+            SuccessOrExit(err);
+
+            err = writer.Finalize(&handle);
+            SuccessOrExit(err);
+
+            // Compact the buffer down to a more reasonably sized packet buffer
+            // if we can.
+            //
+            handle.RightSize();
+
+            mBufferedList.push_back(std::move(handle));
+        }
+
+        err = apData->ExitContainer(outerContainer);
+        SuccessOrExit(err);
+    }
+    else if (aPath.mListOp == ConcreteAttributePath::ListOperation::AppendItem)
+    {
+        System::PacketBufferTLVWriter writer;
+        System::PacketBufferHandle handle = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLengthBytes);
+
+        writer.Init(std::move(handle), false);
+
+        err = writer.CopyElement(TLV::AnonymousTag, *apData);
+        SuccessOrExit(err);
+
+        err = writer.Finalize(&handle);
+        SuccessOrExit(err);
+
+        //
+        // Compact the buffer down to a more reasonably sized packet buffer
+        // if we can.
+        //
+        handle.RightSize();
+
+        mBufferedList.push_back(std::move(handle));
+    }
+
+exit:
+    return err;
+}
+
+CHIP_ERROR BufferedReadCallback::DispatchBufferedData(const ReadClient * apReadClient, const ConcreteAttributePath & aPath,
+                                                      const StatusIB & aStatusIB, bool endOfReport)
+{
+    CHIP_ERROR err;
+
+    if (aPath == mBufferedPath)
+    {
+        //
+        // If we encountered the same list again and it's not the last DataIB, then
+        // we need to continue to buffer up this list's data, so return immediately without dispatching
+        // the existing buffered up contents.
+        //
+        if (!endOfReport)
+        {
+            return CHIP_NO_ERROR;
```

Comment:
you already return directly here. Lets not use both return and goto exit in the same method. Just use the return macros.